internal/pally/labels.go

```diff
@@ -66,3 +69,136 @@ func (d *digester) free() {
 
 // Labels describe the dimensions of a metric.
 type Labels map[string]string
+
+// IsValidName checks whether the supplied string is a valid metric and label
+// name in both Prometheus and Tally.
+//
+// Tally and Prometheus each allow runes that the other doesn't, so Pally can
+// accept only the common subset. For simplicity, we'd also like the rules for
+// metric names and label names to be the same even if that's more restrictive
+// than absolutely necessary.
+//
+// Tally allows anything matching the regexp `^[0-9A-z_\-]+$`. Prometheus
+// allows the regexp `^[A-z_:][0-9A-z_:]*$` for metric names, and
+// `^[A-z_][0-9A-z_]*$` for label names.
+//
+// The common subset is `^[A-z_][0-9A-z_]+$`.
+func IsValidName(s string) bool {
+	if len(s) == 0 {
+		return false
+	}
+
+	switch c := s[0]; {
+	case 'A' <= c && c <= 'Z':
+		break
+	case 'a' <= c && c <= 'z':
+		break
+	case c == '_':
+		break
+	default:
+		return false
+	}
+
+	// Don't incur the expense of ranging over runes, since no multibyte UTF-8
+	// characters are legal.
+	for i := 1; i < len(s); i++ {
+		c := s[i]
+		switch {
+		case '0' <= c && c <= '9':
+			continue
+		case 'A' <= c && c <= 'Z':
+			continue
+		case 'a' <= c && c <= 'z':
+			continue
+		case c == '_':
+			continue
+		default:
+			return false
+		}
+	}
+
+	return true
+}
+
+// IsValidLabelValue checks whether the supplied string is a valid label value
+// in both Prometheus and Tally.
+//
+// Tally allows label values that match the regexp `^[0-9A-z_\-.]+$`.
```

Comment:
is there an argument for why we're not using a regex to validate these strings? (easier to read? Efficiency?)